### PR TITLE
[WIP] Fix Issue 18718 - ICE in dmd/traits.d:417

### PR DESCRIPTION
**Pull request by wilzbach**
I know that this isn't "the right fix", but I didn't know DMD protects itself from self-recursion.
Are there any precendents I can follow?

### CODE REVIEW

File: src/dmd/traits.d [hunk 1]
```diff
@@ -1332,7 +1332,17 @@ extern (C++) Expression semanticTraits(TraitsExp e, Scope* sc)
             return 0;
         }
 
-        ScopeDsymbol._foreach(sc, sds.members, &pushIdentsDg);
+        __gshared int counter;
+        if (counter > 1000)
```
Comment (bausshf) [inline src/dmd/traits.d]: How is this counter calculated, like why is it exactly when it's above 1000?

Comment (wilzbach) [reply src/dmd/traits.d]: Well, it's my noobish attempt at fixing this stack overflow and completely arbitrary. This function (`semanticTraits`) recursively calls itself and I bet there's a more sophisticated way in DMD to protect itself against such recursive overflows, but I couldn't find one.

Comment (RazvanN7) [reply src/dmd/traits.d]: Actually, there isn't. Look how CTFE detects infinite recursion:
https://github.com/dlang/dmd/blob/master/src/dmd/dinterpret.d#L923
where CTFE_RECURSION_LIMIT is ... 1000.  

Or template instantiation: https://github.com/dlang/dmd/blob/master/src/dmd/dsymbolsem.d#L5640 

I guess this fix is fine for now, until a better mechanism is found.

**Review by RazvanN7 — APPROVED** [review]

File: src/dmd/traits.d (hunk 1 above)
Comment (JinShil) [reply src/dmd/traits.d]: Would it make sense to leverage the same limit, rather than create a new one?

File: src/dmd/traits.d [hunk 2]
```diff
@@ -1332,7 +1332,17 @@ extern (C++) Expression semanticTraits(TraitsExp e, Scope* sc)
...
+        {
+            e.error("Maximal recursion depth hit");
```
Comment (JinShil) [inline src/dmd/traits.d]: I recommend "recursion limit exceeded" to remain consistent with the wording used at https://github.com/dlang/dmd/blob/master/src/dmd/dinterpret.d#L927

File: src/dmd/traits.d (hunk 1 above)
Comment (RazvanN7) [reply src/dmd/traits.d]: I guess we could add a symbol recursionLimit which can be used throughout the compiler

File: src/dmd/traits.d (hunk 2 above)
Comment (stefan-koch-sociomantic) [reply src/dmd/traits.d]: it can't hurt to make it explicit.

File: src/dmd/traits.d [hunk 3]
```diff
@@ -1332,7 +1332,17 @@ extern (C++) Expression semanticTraits(TraitsExp e, Scope* sc)
...
+        }
+        else
+        {
+            counter++;
+            ScopeDsymbol._foreach(sc, sds.members, &pushIdentsDg);
+        }
```
Comment (stefan-koch-sociomantic) [inline src/dmd/traits.d]: decrease counter here, otherwise you are counting static_foreach frequency and not just recustion-depth.

**Comment by jacob-carlborg** [comment]
@wilzbach ping, what's the status of this?

**Comment by wilzbach** [comment]
Superseded by https://github.com/dlang/dmd/pull/8213

**Pull request closed without merging**